scripts/run-arm

```diff
@@ -45,154 +45,187 @@ machine=unknown
-	cpu=cortex-m4
-	;;
-    *thumb_v7e_m_fp_hard*)
-	cpu=cortex-m4
-	;;
-    *thumb_v7e_m_fp_softfp*)
-	cpu=cortex-m4
-	;;
-
-    # v7e-m dp - Cortex-M7
-    *thumb_v7e_m_dp_hard*)
-	cpu=cortex-m7
-	;;
-    *thumb_v7e_m_dp_softfp*)
-	cpu=cortex-m7
-	;;
-
-    # v8-a - unknown
-    *thumb_v8_a*)
-	cpu=unknown
-	;;
-
-    # v8.1 - unknown
-    *thumb_v8_1*)
-	cpu=unknown
-	;;
-
-    # v8-m with single-precision FPU - Cortex-M33
-    *thumb_v8_m_base_nofp*)
-	cpu=cortex-m33
-	;;
-    *thumb_v8_m_main_fp_softfp*)
-	cpu=cortex-m33
-	;;
-    *thumb_v8_m_main_dp_softfp*)
-	;;
-    *thumb_v8_m_main_fp_hard*)
-	cpu=cortex-m33
-	;;
-    *thumb_v8_m_main_nofp*)
-	cpu=cortex-m33
-	;;
-
-    # v8-m with double-preicision FPU - none
-    *thumb_v8_m_main_dp_hard*)
-	;;
-
-    *thumb_v8_1_m_main_mve_hard*)
-    cpu=cortex-m55
-    ;;
-
-    # v7 - Cortex-A7
-    *thumb_v7_nofp*)
-	cpu=cortex-a7
-	;;
-
-    # v7 with FPU - Cortex-A9
-    *thumb_v7_fp_hard*)
-	cpu=cortex-a8
-	;;
-    *thumb_v7_fp_softfp*)
-	cpu=cortex-a8
-	;;
+cpu_arch=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_CPU_arch:/ { print $2 }'`
+cpu_profile=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_CPU_arch_profile:/ { print $2 }'`
+fp_arch=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_FP_arch/ { print $2 }'`
+fp_use=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_ABI_HardFP_use/ { print $2 }'`
 
-    # v4t - ti925t
-    *)
-	# ARMv4t is the default
-	cpu=ti925t
-	;;
+case "$cpu_arch"/"$cpu_profile" in
+    v4T/)
+        cpu=ti925t
+        ;;
+    v5TE/)
+        cpu=arm926
+        ;;
+    v6S-M/Microcontroller)
+        cpu=cortex-m0
+        ;;
+    v7/Application|v7/)
+        case "$fp_arch" in
+            VFPv3-D16)
+                cpu=cortex-a8
+                ;;
+            *)
+                cpu=cortex-a7
+                ;;
+        esac
+        ;;        
+    v7/Microcontroller)
+        cpu=cortex-m3
+        ;;
+    v7/Realtime)
+        cpu=cortex-r5f
+        ;;
+    v7E-M/Microcontroller)
+        cpu=cortex-m7
+        case "$fp_arch" in
+            FPv5/FP-D16)
+                case "$fp_use" in
+                    SP)
+                        cpu=cortex-m4
+                        ;;
+                esac
+                ;;
+            *)
+                cpu=cortex-m4
+                ;;
+        esac
+        ;;
+    
+    v8-M.*/Microcontroller)
+        case "$fp_arch" in
+            FPv5/FP-D16)
+                case "$fp_use" in
+                    SP)
+                        cpu=cortex-m33
+                        ;;
+                    *)
+                        cpu=cortex-m55
+                        ;;
+                esac
+                ;;
+            *)
+                cpu=cortex-m33
+                ;;
+        esac
+        ;;
+    v8.1-M.mainline/Microcontroller)
+        case "$fp_arch" in
+            FPv5/FP-D16)
+                case "$fp_use" in
+                    SP)
+                        cpu=cortex-m55
+                        ;;
```

Comment:
I can see cortex-m55 is used for double precision case in "v8-M.*". Cannot it be used here for both single and double precision?

Reply:
One of the goals here is to make sure the code doesn't use features not found in the target architecture, so limiting the core to something that more precisely fits the compiler settings is why I have stuff like this. In this case, I believe the m33 is a v8 core with SP float while the m55 is a v8.1 core with SP/DP float. Or I could just be wrong :-)

Reply:
I agree it's better to use cortex-m33 for SP in 8m case. I just wonder why, for the 8.1m case, m55 is used only for SP and DP is left unhandled. I thought m55 should be able to handle DP case, too.
Anyway, gcc 12.2 does not provide a multilib variant for v8.1m+fp.dp. As there would be no test to confirm the configuration works, maybe it's better to leave it as is for now.

Reply:
Oh, now I see your point. It looks like cortex-m55 could handle all of the v8.1m cases, even those not currently part of the gcc multilib configuration? Changing that would simplify the script, which seems like a good idea.